### PR TITLE
Support for auth to registry via client certificates in PKCS11

### LINKED ISSUE #38849: PKCS#11 Registry Authentication
A number of organisations have adopted PIV smartcard devices, such as Yubikeys for TLS client authentication. It would be beneficial to allow the use of PKCS#11 interfaces when connecting to a registry. 

Currently Docker allows configuration of the client certificate and key in the `/etc/docker/certs.d/` directory. This requires the certificates and keys to be in PEM format. 

Possible approaches:
1. Specify `MODULE_PATH` and object identifiers in config.json 
2. Support a new file type in `/etc/docker/certs.d/` that contains `MODULE_PATH` and PKCS11 object identifiers
3. Allow setting of `MODULE_PATH` and object identifiers as ENV vars.

Some gotchas 
1. Capturing input for user pin?
2. Ensuring permissions of `MODULE_PATH` are valid to prevent privilege escalation. 

Will have a bash at implementing this, but would appreciate some discussion/guidance on the preferred approach. 

Most of the changes will need to be made to: https://github.com/moby/moby/blob/master/registry/registry.go#L57

### PR DESCRIPTION
**Pull request by jeffallen**
If your registry requires TLS client certificates, and your certificate is stored in a Yubikey, you cannot use the existing support for reading private keys out of the certs directory.

This change only works when dockerd is built to link dynamically.

Fixes #38849

**- Description for the changelog**

If compiled with `make dynbinary`, dockerd will read various environment variables to control access to a PKCS11 device (i.e. Yubikey, etc) to get the client certificate.

```markdown changelog
If compiled with `make dynbinary`, dockerd will read various environment variables to configure access to a PKCS11 device (i.e. Yubikey, etc) to get the client certificate used to talk to registries.
```

### CODE REVIEW

File: vendor.mod [hunk 1]
```diff
@@ -191,6 +193,7 @@ require (
 	github.com/spdx/tools-golang v0.5.3 // indirect
 	github.com/stretchr/testify v1.8.4 // indirect
 	github.com/syndtr/gocapability v0.0.0-20200815063812-42c35b437635 // indirect
+	github.com/thales-e-security/pool v0.0.2 // indirect
```
Comment (thaJeztah) [inline vendor.mod:196]: This looks to be some un-blessed fork of https://github.com/vitessio/vitess; last commit is 4 Years ago, and last code-commit 5 Years ago 😬 https://github.com/thales-e-security/pool
Screenshot 2024-05-30 at 11 26 43

File: vendor.mod [hunk 2]
```diff
@@ -17,6 +17,7 @@ require (
 	github.com/Microsoft/go-winio v0.6.2
 	github.com/Microsoft/hcsshim v0.11.4
 	github.com/RackSec/srslog v0.0.0-20180709174129-a4725f04ec91
+	github.com/ThalesIgnite/crypto11 v1.2.5
```
Comment (thaJeztah) [inline vendor.mod:20]: This repository also looks dormant; last commit 3 Years ago, which doesn't bring a lot of confidence; https://github.com/ThalesGroup/crypto11

File: registry/registry.go [hunk 3]
```diff
@@ -36,6 +40,71 @@ func newTLSConfig(hostname string, isSecure bool) (*tls.Config, error) {
...
+	//
+	//   PKCS11_LABEL=my.cert.name
+	//   PKCS11_PIN=112233
+	//   PKCS11_MODULE=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so
+	//
+	// You can use only one of PKCS11_SERIAL, PKCS11_SLOT or
+	// PKCS11_LABEL.
+	//
+	// Then `systemctl restart docker.service` to make it take effect.	
+	pkcs11_module := os.Getenv("PKCS11_MODULE")
```
Comment (thaJeztah) [inline registry/registry.go:64]: I'm also not sure if this is the right location to do this; this code is all on the daemon side, 


- to make this all work, we should also look at containerd code, as well as BuildKit, as all of them may be interacting with registries
- should authentication here happen on the daemon side, or something to be passed from the client?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Leaving a "request changes" because this may need further discussion.

File: registry/registry.go (hunk 3 above)
Comment (jeffallen) [reply registry/registry.go:64]: I agree that you would normally expect this to come from the client (same for the private key for client authentication, which is also read here and arguably should not be). Doing that would be a major change, which I don't have the appetite to take on, and I don't think the project would accept anyway. (There has already once before been an effort to *remove* the dependency on miekg/pkcs11, in fact.)

While developing this, I did find myself single stepping far into containerd, but finally I found that I could get `docker pull` and `docker build` working with only these changes, and that's what I needed to get unblocked on my own problem.

I decided to share this now to have this exact conversation.

File: vendor.mod (hunk 2 above)
Comment (jeffallen) [reply vendor.mod:20]: I did not notice that, if we decide to go forward with this PR, I will look into finding a better choice here. This package was chosen by a colleague for another place we use PKCS#11 hardware tokens, I'll ask her about it.

**Comment by jeffallen** [comment]
Close because there seems to be no interest in this.

**Pull request closed without merging**